Lilu/Headers/kern_util.hpp

```diff
@@ -944,4 +945,285 @@ class evector_base {
 template <typename T, void (*deleter)(T)=emptyDeleter<T>>
 class evector : public evector_base<typename remove_reference<T>::type, T, deleter> { };
 
+/**
+ *  Represents a circular buffer protected by a recursive mutex lock
+ */
+template <typename T>
+struct CircularBuffer {
+private:
+	/**
+	 *  The internal storage
+	 */
+	T *storage;
+	
+	/**
+	 *  The buffer capacity
+	 */
+	IOItemCount size;
+	
+	/**
+	 *  The current index for the next read operation
+	 */
+	IOItemCount indexr;
+	
+	/**
+	 *  The current index for the next write operation
+	 */
+	IOItemCount indexw;
+	
+	/**
+	 *  The current number of elements in the buffer
+	 */
+	IOItemCount count;
+	
+	/**
+	 *  The recursive mutex lock that protects the buffer
+	 */
+	IORecursiveLock *lock;
+	
+public:
+	/**
+	 *  Initialize a circular buffer
+	 *
+	 *  @param buffer A non-null storage buffer
+	 *  @param capacity The total number of elements
+	 *  @return `true` on success, `false` otherwise.
+	 *  @warning The caller is responsbile for managing the lifecycle of the given storage buffer.
+	 */
+	bool init(T *buffer, IOItemCount capacity) {
+		storage = buffer;
+		size = capacity;
+		indexr = 0;
+		indexw = 0;
+		count = 0;
+		lock = IORecursiveLockAlloc();
+		return lock != nullptr;
+	}
+	
+	/**
+	 *  Initialize a circular buffer
+	 *
+	 *  @param storage A storage buffer
+	 *  @return `true` on success, `false` otherwise.
+	 *  @warning The caller is responsbile for managing the lifecycle of the given storage buffer.
+	 */
+	template <size_t N>
+	bool init(T (&storage)[N]) {
+		return init(storage, N);
+	}
+	
+	/**
+	 *  Deinitialize the circular buffer
+	 */
+	void deinit() {
+		IORecursiveLockFree(lock);
+	}
+	
+	/**
+	 *  Create a circular buffer with the given capacity
+	 *
+	 *  @param size The total number of elements
+	 *  @return A non-null instance on success, `nullptr` if no memory.
+	 *  @warning The caller must invoke `CircularBuffer::destory()` to release the returned buffer.
+	 */
+	static CircularBuffer<T> *withCapacity(IOItemCount size) {
+		auto storage = IONewZero(T, size);
+		if (storage == nullptr)
+			return nullptr;
+		
+		auto instance = new CircularBuffer<T>();
+		if (instance == nullptr)
+			goto error1;
+		
+		if (!instance->init(storage, size))
+			goto error2;
+		
+		return instance;
+		
+	error2:
+		delete instance;
+		
+	error1:
+		IODelete(storage, T, size);
+		return nullptr;
+	}
+	
+	/**
+	 *  Destory the given circular buffer
```

Comment:
I believe it is a typo from destroy. But probably the most common convention in Lilu is deleter. It also allows storing elements of this type in evector.